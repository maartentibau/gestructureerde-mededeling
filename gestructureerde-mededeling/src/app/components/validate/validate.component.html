<ogm-number [ogm]="ogm$ | async" [isValid]="isValid$ | async"></ogm-number>

<ogm-input
  (ogmInputChange)="ogmInputChangeHandler($event)"
  [validate]="true"
  placeholderMessage="vul de 12 cijfers in van uw gestructureerde mededeling"
></ogm-input>
