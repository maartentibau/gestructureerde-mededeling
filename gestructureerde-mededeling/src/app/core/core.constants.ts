export const DEFAULT_TITLE = `Gestructureerde Mededeling - OGM`;
