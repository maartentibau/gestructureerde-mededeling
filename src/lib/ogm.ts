export function ogm(): string {
  return 'ogm';
}
