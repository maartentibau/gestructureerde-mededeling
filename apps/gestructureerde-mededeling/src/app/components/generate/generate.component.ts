import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
import { MatSnackBar } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmService } from '../../core/services/ogm.service';
import { OgmData } from '../../core/ogm.model';

@Component({
  selector: 'ogm-generate',
  templateUrl: './generate.component.html',
  styleUrls: ['./generate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GenerateComponent implements OnInit {
  ogm: OgmData | null | undefined;

  constructor(private ogmService: OgmService, private snackBar: MatSnackBar, private title: Title) {
    this.title.setTitle(`${DEFAULT_TITLE} - Genereer een willekeurig gestructureerde mededeling`);
  }

  ngOnInit() {
    this.refreshClickHandler();
  }

  refreshClickHandler() {
    const ogmNumber = this.ogmService.generate();
    const ogmNumberFormat = this.ogmService.format(ogmNumber);

    this.ogm = { number: ogmNumber, numberFormat: ogmNumberFormat };
  }

  copyNumberClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  copyOgmClickHandler(message: string) {
    this.showSnackbarMessage(message);
  }

  showSnackbarMessage(message?: string) {
    if (message) {
      this.snackBar.open(message, '', {
        panelClass: 'center-clipboard-text',
        verticalPosition: 'top',
        duration: 750,
      });
    }
  }
}
