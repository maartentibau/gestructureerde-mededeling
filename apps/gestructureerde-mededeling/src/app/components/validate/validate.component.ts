import { ChangeDetectionStrategy, Component } from '@angular/core';
import { BehaviorSubject, Subject } from 'rxjs';
import { OgmService } from '../../core/services/ogm.service';
import { OgmInputChange } from '../../core/components/input/input.component';

@Component({
  selector: 'ogm-validate',
  templateUrl: './validate.component.html',
  styleUrls: ['./validate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ValidateComponent {
  readonly ogm$: BehaviorSubject<string>;
  readonly isValid$: Subject<boolean>;

  constructor(private ogmService: OgmService) {
    this.ogm$ = new BehaviorSubject<string>(this.ogmService.init());
    this.isValid$ = new Subject<boolean>();
  }

  ogmInputChangeHandler({ ogm, isValid }: OgmInputChange) {
    this.ogm$.next(ogm);
    this.isValid$.next(isValid);
  }
}
