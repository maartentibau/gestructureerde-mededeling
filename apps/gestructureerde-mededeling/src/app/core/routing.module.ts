import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { CreateComponent } from '../components/create/create.component';
import { GenerateComponent } from '../components/generate/generate.component';
import { ValidateComponent } from '../components/validate/validate.component';

const routes: Routes = [
  { path: 'create', component: CreateComponent },
  { path: 'generate', component: GenerateComponent },
  { path: 'validate', component: ValidateComponent },
  { path: '', redirectTo: '/generate', pathMatch: 'full' },
  { path: '**', component: GenerateComponent },
];

@NgModule({
  imports: [RouterModule.forRoot(routes, { relativeLinkResolution: 'legacy' })],
  exports: [RouterModule],
})
export class RoutingModule {}
