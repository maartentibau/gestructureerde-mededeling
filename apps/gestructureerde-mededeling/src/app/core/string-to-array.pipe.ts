import { Pipe, PipeTransform } from '@angular/core';

@Pipe({
  name: 'stringToArray',
})
export class StringToArrayPipe implements PipeTransform {
  transform(value: string | null | undefined): string[] {
    if ((value && typeof value === 'string') || typeof value === 'number') {
      return Array.from(String(value));
    }

    return [];
  }
}
