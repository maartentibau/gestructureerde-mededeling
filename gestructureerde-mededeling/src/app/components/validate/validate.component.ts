import { AsyncPipe } from '@angular/common';
import { ChangeDetectionStrategy, Component, inject } from '@angular/core';
import { Title } from '@angular/platform-browser';
import { BehaviorSubject, Subject } from 'rxjs';
import { InputComponent, OgmInputChange } from '../../core/components/input/input.component';
import { NumberComponent } from '../../core/components/number/number.component';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { OgmService } from '../../core/services/ogm.service';

@Component({
  standalone: true,
  imports: [NumberComponent, InputComponent, AsyncPipe],
  selector: 'ogm-validate',
  templateUrl: './validate.component.html',
  styleUrls: ['./validate.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class ValidateComponent {
  #ogmService: OgmService = inject(OgmService);
  #title: Title = inject(Title);

  readonly ogm$: BehaviorSubject<string>;
  readonly isValid$: Subject<boolean | null>;

  constructor() {
    this.#title.setTitle(`${DEFAULT_TITLE} - Controleer een gestructureerde mededeling`);

    this.ogm$ = new BehaviorSubject<string>(this.#ogmService.init());
    this.isValid$ = new Subject<boolean | null>();
  }

  ogmInputChangeHandler({ ogm, isValid }: OgmInputChange) {
    this.ogm$.next(ogm);
    this.isValid$.next(isValid);
  }
}
