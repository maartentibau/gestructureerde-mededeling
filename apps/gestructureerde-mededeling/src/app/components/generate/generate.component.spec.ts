import { Component, Input } from '@angular/core';
import { async, ComponentFixture, TestBed } from '@angular/core/testing';
import { MatSnackBar, MatSnackBarConfig } from '@angular/material/snack-bar';
import { Title } from '@angular/platform-browser';
import { DEFAULT_TITLE } from '../../core/core.constants';
import { MaterialModule } from '../../core/material.module';
import { OgmData } from '../../core/ogm.model';
import { OgmService } from '../../core/services/ogm.service';
import { ValidateComponent } from '../validate/validate.component';

import { GenerateComponent } from './generate.component';

@Component({
  selector: 'ogm-number',
  template: `
    <div>{{ ogm | json }}</div>
    <div>{{ isValid | json }}</div>
  `,
})
class MockNumberComponent {
  @Input() ogm: string;
  @Input() isValid: boolean;
}

@Component({
  selector: 'ogm-input',
  template: `
    <div>{{ validate | json }}</div>
    <div>{{ placeholderMessage | json }}</div>
  `,
})
class MockInputComponent {
  @Input() validate: boolean;
  @Input() placeholderMessage: string;
}

@Component({
  selector: 'ogm-controls',
  template: `
    <div>{{ refresh | json }}</div>
    <div>{{ copyNumber | json }}</div>
    <div>{{ copyOgm | json }}</div>
    <div>{{ ogm | json }}</div>
  `,
})
class MockControlsComponent {
  @Input() refresh: boolean;
  @Input() copyNumber: boolean;
  @Input() copyOgm: boolean;
  @Input() ogm: OgmData;
}

describe('GenerateComponent', () => {
  let component: GenerateComponent;
  let fixture: ComponentFixture<GenerateComponent>;
  let snackBar: MatSnackBar;
  let ogmService: OgmService;
  let titleService: Title;

  const number: string = '123456789012';
  const numberFormat: string = '+++123/4567/89012+++';

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [MaterialModule],
      declarations: [GenerateComponent, MockNumberComponent, MockInputComponent, MockControlsComponent],
      providers: [OgmService, MatSnackBar],
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(GenerateComponent);
    component = fixture.componentInstance;

    snackBar = TestBed.inject(MatSnackBar);
    ogmService = TestBed.inject(OgmService);
    titleService = TestBed.inject(Title);

    jest.spyOn(ogmService, 'generate').mockReturnValueOnce(number);
    jest.spyOn(ogmService, 'format').mockReturnValueOnce(numberFormat);
    jest.spyOn(snackBar, 'open');
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  describe('rendering component', () => {
    it('should match snapshot', () => {
      // act
      fixture.detectChanges();

      // check
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('set the correct HTML title', () => {
      // prepare
      const title: string = `${DEFAULT_TITLE} - Genereer een willekeurig gestructureerde mededeling`;
      jest.spyOn(titleService, 'setTitle');

      // act
      TestBed.createComponent(GenerateComponent);

      // check
      expect(titleService.setTitle).toHaveBeenCalledWith(title);
    });
  });

  describe('ngOnInit', () => {
    it('should call refresh method', () => {
      // prepare
      jest.spyOn(component, 'refreshClickHandler');

      // act
      component.ngOnInit();

      // check
      expect(component.refreshClickHandler).toHaveBeenCalled();
    });
  });

  describe('refreshClickHandler', () => {
    it('should refresh the current OGM number', () => {
      // prepare
      const expectedResult = {
        number,
        numberFormat,
      };

      // act
      component.refreshClickHandler();

      // check
      expect(ogmService.generate).toHaveBeenCalled();
      expect(ogmService.format).toHaveBeenCalled();
      expect(component.ogm).toStrictEqual(expectedResult);
    });
  });

  describe('Copy number or OGM', () => {
    beforeEach(() => {
      jest.spyOn(component, 'showSnackbarMessage');
    });

    describe('copyNumberClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy number!';

        // act
        component.copyNumberClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });

    describe('copyOgmClickHandler', () => {
      it('should the correct copy message when called', () => {
        // prepare
        const message: string = 'Copy OGM!';

        // act
        component.copyOgmClickHandler(message);

        // check
        expect(component.showSnackbarMessage).toHaveBeenCalledWith(message);
      });
    });
  });

  describe('showSnackbarMessage', () => {
    const snackBarConfig: MatSnackBarConfig = {
      panelClass: 'center-clipboard-text',
      verticalPosition: 'top',
      duration: 750,
    };

    it('should call open when a message is provided', () => {
      // prepare
      const message: string = 'Hello World!';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).toHaveBeenCalledWith(message, '', snackBarConfig);
    });

    it('should NOT call open when no message is provided', () => {
      // prepare
      const message: string = '';

      // act
      component.showSnackbarMessage(message);

      // check
      expect(snackBar.open).not.toHaveBeenCalled();
    });
  });
});
