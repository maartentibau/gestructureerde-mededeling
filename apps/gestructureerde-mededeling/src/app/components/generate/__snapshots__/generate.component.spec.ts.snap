// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenerateComponent rendering component should match snapshot 1`] = `
<ogm-generate
  __ngContext__={[Function Number]}
  ogm={[Function Object]}
  ogmService={[Function OgmService]}
  snackBar={[Function MatSnackBar]}
  title={[Function Title]}
>
  <ogm-number>
    <div>
      "+++123/4567/89012+++"
    </div>
    <div>
    </div>
  </ogm-number><ogm-controls>
    <div>
      true
    </div>
    <div>
      true
    </div>
    <div>
      true
    </div>
    <div>
      {
    "number": "123456789012",
    "numberFormat": "+++123/4567/89012+++"
  }
    </div>
  </ogm-controls>
</ogm-generate>
`;
